### DISCUSSION TITLE
glsp server as web worker

### DISCUSSION THREAD

**Discussion by kchobantonov** [eclipse-glsp/glsp#1114] [Q&A] [answered]
Hello, 

Is it feasible to to implement GLSP client and server that both work in the same browser instance (using web worker) and not have a dedicated backend for the server, e.g. the load the diagram data initially, pass it to the web worker and then the glsp client to communicate with the web worker.

Also it if it possible how much effort it is to do that if known.

Thanks

**Comment by tortmayr** [comment, marked as answer]
Hi @kchobantonov,

yes this is definitly feasible. We are currently working on finalizing the 2.0 Release of GLSP which also ships with full support to
run GLSP in a browser-only environment. Once this release is available the effort is more or less the same as developing a classical GLSP application with a node-backend. If you are interested you can find the source code for a (WIP) example [here](https://github.com/eclipse-glsp/glsp-examples/commit/25f7c220bc3f4cf0ce5e5591332032829e526faf) and the deployed example app [here](https://eclipse-glsp.github.io/glsp-examples/).

The main steps are:

- Implement your GLSP server using the components from `@eclipse-glsp/server/browser`.
- Provide a  module for your webworker that launches the server. We provide a `WorkerServerLauncher` component to simplify this step.
  (See also [workflow-live-demo/src/server-worker.ts)](https://github.com/eclipse-glsp/glsp-examples/blob/25f7c220bc3f4cf0ce5e5591332032829e526faf/workflow-live-demo/src/server-worker.ts)
- Setup the main module of your client application (establish a connection to the worker and initialize your diagram)
  (see also [workflow-live-demo/src/main.ts](https://github.com/eclipse-glsp/glsp-examples/blob/25f7c220bc3f4cf0ce5e5591332032829e526faf/workflow-live-demo/src/main.ts))
  

There is only one additional challenge that you have to consider here. The default `SourceModelStorage´ implementations work on top of the node File API. Since this is not available in the browser your have to implement a custom  approach to retrieve your source model.

A real example and more documentation will be provided once the 2.0 release is published.
